packages/common/context/src/resource.ts

```diff
@@ -0,0 +1,131 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import '@dxos/util';
+import { Context } from './context';
+
+export interface Lifecycle {
+  open?(ctx?: Context): Promise<any> | any;
```

Comment:
why are methods optional? 🤔

Reply:
To allow typescript to structurally match types that aren't `Resource`. For example `openInContext` below could work on any time that has open and/or close.